src/core/ext/gcp/metadata_query.cc

```diff
@@ -0,0 +1,113 @@
+// limitations under the License.
+//
+//
+
+#include <grpc/support/port_platform.h>
+
+#include "src/core/ext/gcp/metadata_query.h"
+
+#include <string.h>
+
+#include <memory>
+#include <utility>
+
+#include "absl/status/statusor.h"
+#include "absl/strings/string_view.h"
+
+#include <grpc/grpc.h>
+#include <grpc/grpc_security.h>
+#include <grpc/support/log.h>
+
+#include "src/core/lib/gprpp/ref_counted_ptr.h"
+#include "src/core/lib/gprpp/status_helper.h"
+#include "src/core/lib/gprpp/time.h"
+#include "src/core/lib/http/httpcli.h"
+#include "src/core/lib/security/credentials/credentials.h"
+#include "src/core/lib/uri/uri_parser.h"
+
+namespace grpc_core {
+
+constexpr const char MetadataQuery::kZoneAttribute[] =
+    "/computeMetadata/v1/instance/zone";
+constexpr const char MetadataQuery::kClusterNameAttribute[] =
+    "/computeMetadata/v1/instance/attributes/cluster-name";
+constexpr const char MetadataQuery::kRegionAttribute[] =
+    "/computeMetadata/v1/instance/region";
+constexpr const char MetadataQuery::kInstanceIdAttribute[] =
+    "/computeMetadata/v1/instance/id";
+constexpr const char MetadataQuery::kIPv6Attribute[] =
+    "/computeMetadata/v1/instance/network-interfaces/0/ipv6s";
+
+MetadataQuery::MetadataQuery(
+    std::string attribute, grpc_polling_entity* pollent,
+    absl::AnyInvocable<void(std::string /* attribute */,
+                            std::string /* result */)>
+        callback)
+    : attribute_(std::move(attribute)), callback_(std::move(callback)) {
+  GRPC_CLOSURE_INIT(&on_done_, OnDone, this, nullptr);
+  auto uri =
+      grpc_core::URI::Create("http", "metadata.google.internal.", attribute_,
+                             {} /* query params */, "" /* fragment */);
+  GPR_ASSERT(uri.ok());  // params are hardcoded
+  grpc_http_request request;
+  memset(&request, 0, sizeof(grpc_http_request));
+  grpc_http_header header = {const_cast<char*>("Metadata-Flavor"),
+                             const_cast<char*>("Google")};
+  request.hdr_count = 1;
+  request.hdrs = &header;
+  // The http call is local. If it takes more than one sec, it is probably not
+  // on GCP.
+  auto http_request = grpc_core::HttpRequest::Get(
+      std::move(*uri), nullptr /* channel args */, pollent, &request,
+      grpc_core::Timestamp::Now() + grpc_core::Duration::Seconds(1), &on_done_,
+      &response_,
+      grpc_core::RefCountedPtr<grpc_channel_credentials>(
+          grpc_insecure_credentials_create()));
+  http_request->Start();
+}
+
+MetadataQuery::~MetadataQuery() { grpc_http_response_destroy(&response_); }
+
+void MetadataQuery::OnDone(void* arg, absl::Status error) {
+  auto* self = static_cast<MetadataQuery*>(arg);
+  std::string result;
+  if (!error.ok()) {
+    gpr_log(GPR_ERROR, "MetadataServer Query failed for %s: %s",
+            self->attribute_.c_str(), grpc_core::StatusToString(error).c_str());
+  } else if (self->response_.status != 200) {
+    gpr_log(GPR_ERROR,
```

Comment:
We shouldn't log ERROR here, since there are some queries that are performed as "probes" for which we may or may not get a 200 response, but we can handle both cases.

E.g. the google-c2p resolver queries the MD server to see if the VM has an IPv6 address, to decide whether to prefer v4 or v6.

If we want to add extra error message context, suggest adding it to the error and letting the caller decide to log or not.

Reply:
Done!